plugins/base/base.go

```diff
@@ -0,0 +1,35 @@
+package shared
+
+import (
+	"github.com/hashicorp/nomad/plugins/shared/hclspec"
+)
+
+// BasePlugin is the interface that all Nomad plugins must support.
+type BasePlugin interface {
+	// PluginInfo describes the type and version of a plugin.
+	PluginInfo() (*PluginInfoResponse, error)
+
+	// ConfigSchema returns the schema for parsing the plugins configuration.
+	ConfigSchema() (*hclspec.Spec, error)
+
+	// SetConfig is used to set the configuration by passing a MessagePack
+	// encoding of it.
+	SetConfig(data []byte) error
+}
+
+// PluginInfoResponse returns basic information about the plugin such that Nomad
+// can decide whether to load the plugin or not.
+type PluginInfoResponse struct {
+	// Type returns the plugins type
+	Type string
+
+	// PluginApiVersion returns the version of the Nomad plugin API it is built
+	// against.
+	PluginApiVersion string
```

Comment:
Do we want to model these versions as `github.com/coreos/go-semver/semver.Version` ptrs? Or may be add Getters to the struct that return that type? Just pondering if we should do some sort of validation on these version strings since they'll later be used for load ordering?

Reply:
We will use this to parse as semver: https://github.com/hashicorp/go-version
Simpler to keep using this library and expose a primitive type for all languages over protobuf.

Reply:
I agree over protobuf. But a helper here in the go interface layer might be nice. Don't want to bikeshed this.